dra-evolution/pkg/api/capacity_types.go

```diff
@@ -1,170 +1,205 @@
+	//
+	// If we had transactions in the apiserver, we could combine a status
+	// update of the ResourcePool with a status update of the claim.  But
+	// we don't and adding it [is
+	// hard](https://kubernetes.slack.com/archives/C0EG7JC6T/p1714373064352099).
+	// Without transactions, there will be a risk of leaking resources
+	// and/or races around freeing leaked resources. This would be great to
+	// have in 1.31 because requiring schedulers to use the ResourcePool
+	// status for allocation will be a change of behavior, but it'll be
+	// hard to design and implement in time.
 }
 
-// Device is used to track individual devices in a pool.
-type Device struct {
-	// Name is a driver-specific identifier for the device.
-	// +required
-	Name string `json:"name"`
-
-	// Attributes contain additional metadata that can be used in
-	// constraints. If an attribute name overlaps with the pool attribute,
-	// the device attribute takes precedence.
+// ResourceModel must have one and only one field set.
+type ResourceModel struct {
+	// NamedDevices describes available devices by listing them.
 	//
 	// +optional
-	Attributes []Attribute `json:"attributes,omitempty"`
+	NamedDevices *NamedDevices `json:"namedDevices,omitempty" protobuf:"bytes,1,opt,name=namedResources"`
+}
 
-	// SharedResourcesConsumed contains the pooled resources that are
-	// consumed when this device is allocated.
+// NamedDevices is used in ResourceModel.
+type NamedDevices struct {
+	// CommonAttributes defines the set of attributes common to all devices in this ResourcePool.
+	// The name of each attribute must be unique.
 	//
+	// +listType=atomic
 	// +optional
-	SharedResourcesConsumed map[string]resource.Quantity `json:"sharedResourcesConsumed,omitempty"`
+	CommonAttributes []DeviceAttribute `json:"commonAttributes" protobuf:"bytes,2,name=commonAttributes"`
 
-	// ClaimResourcesProvided allows the definition of per-device resources
-	// that can be allocated in a manner similar to standard Kubernetes
-	// resources.
+	// A set of shared resources that get consumed as devices in this
+	// ResourcePool get allocated out.
+	SharedResources []DeviceResource `json:"sharedResources" protobuf:"bytes,3,name=sharedResources"`
+
+	// The list of all devices currently available.
 	//
-	// +optional
-	ClaimResourcesProvided []ResourceCapacity `json:"claimResourcesProvided,omitempty"`
+	// +listType=atomic
+	Devices []Device `json:"devices" protobuf:"bytes,1,name=instances"`
 }
 
-type ResourceCapacity struct {
-	// Name is the resource name/type.
-	// +required
-	Name string `json:"name"`
-
-	// Capacity is the total capacity of the named resource.
-	// +required
-	Capacity resource.Quantity `json:"capacity"`
+// Device represents one individual hardware instance that can be selected based
+// on its attributes.
+type Device struct {
+	// Name is unique identifier among all devices managed by
+	// the driver on the node. It must be a DNS subdomain.
```

Comment:
We should have some validation for strings. "Anything goes, with arbitrary length" isn't good enough for the Kubernetes API.

Picking "DNS subdomain" was arbitrary. We can also allow a "DNS domain". I wouldn't go further, because reserving special characters like slash for use by code which needs to work with these names can be useful.

For example, `<driver name>/<node name>/<device name>` could be a useful single-string ID, but it only works if neither driver name nor device name are allowed to contain a slash.

Reply:
> So <driver name>/<device name> would be the corresponding "unique ID string" for non-local devices.

Is that right?  I can imagine a rack-switched device where different pools on nodes have different "domains" of the device.  Ie "this device is owned by driver foo and can be used on any node where rack == A" vs racxk B, C, etc.  Do we _require_ the device name to be unique across racks?

Reply:
The driver doesn't need to use a "readable" name. It could be a UUID, or the driver could embed the rack number. We just need to articulate clearly what kind of uniqueness is expected from the driver.

Reply:
I guess I am just saying "unique within the set of all nodes which can use the device" is sufficient.  "unique within all nodes" is an even stricter form of that.  Do we need the stricter form?

Reply:
"unique within the set of all nodes which can use the device" means that the tuple `(<driver name>, <node selector>, <device name>)` becomes the unique identifier for the device. That's not going to work in e.g. a hash. I think we need the stricter form.

Reply:
If this is getting too fuzzy, then we can also just say "each device must have a UUID".

Reply:
I think having a device UUID is probably the right thing to do. It would need to be preserved across kubelet, driver and node reboots. Maybe not node reboots if we evict everything in that case?

If it needs to survive all of that, it likely needs to come from the hardware device itself or at least be persisted with some way to map it back to the physical card. But you will then have a problem if you expect to be able to hotswap a failed device...

Reply:
> Is there some need to hash this to a unique value?

I meant using it as key in a map. Sorry, brain not working properly anymore.

> But you will then have a problem if you expect to be able to hotswap a failed device...

That's a good point. A UUID might be too specific. Also, having to print it in debug messages instead of the (probably) more readable name makes the debug output less readable.

How about this:
- For node-local resources, `<driver name>/<node name>/<device name>` must be unique.
- For non-node-local resources, drivers may supply a `<device pool name>` together with the `nodeSelector` and then `<driver name>/<device pool name>/<device name>` must be unique. It's okay for `<device pool name>` to be empty if the `<device name>` is already unique in the cluster.

Reply:
I know GKE, for example, would take issue with forcing one to use a UUID as the name.

With the existing device plugin, GKE purposefully uses an index rather than a UUID to represent the "name" of their devices. This is necessary because node reboots on GKE do not guarantee that the exact same GPU hardware gets mapped into the rebooted node (even though the node-name stays the same). So long as the node is identically "shaped", using an index for the GPUs guarantees that the pod will still run even if the actual UUID of the GPUs are different.

Reply:
Oo, good to know. Fun to learn about GKE from Kevin :laughing: !

Reply:
Let's go with DNS label.

I've removed the node selector, so we don't need to worry about that now.